### PR TITLE
Add object detection test case plots

### PR DESCRIPTION
**Pull request by marko-polo-cheno**
PR + F1 plots

### CODE REVIEW

File: kolena/workflow/metrics/_plots.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+def _compute_sklearn_arrays(
+    all_matches: List[Union[MulticlassInferenceMatches, InferenceMatches]],
+) -> Tuple[List[int], List[int], Dict[str, Tuple[List[int], List[int]]]]:
+    arrays_by_label: defaultdict[str, Tuple[List[int], List[int]]] = defaultdict(lambda: ([], []))
+    y_true: List[int] = []
+    y_score: List[int] = []
+    for image_bbox_matches in all_matches:
+        for _, bbox_inf in image_bbox_matches.matched:  # TN (if above threshold)
+            arrays_by_label[bbox_inf.label][0].append(1)
+            arrays_by_label[bbox_inf.label][1].append(bbox_inf.score)
```
Comment (munkyshi) [inline kolena/workflow/metrics/_plots.py]: nit: The structure and naming for `arrays_by_label` makes it hard to read what's going on (What is `arrays`, what is the indexing in the tuple?).

Thoughts on:
```
y_true_by_label[bbox_inf.label].append(1)
y_score_by_label[bbox_inf.label].append(bbox_inf.score)
```

File: kolena/workflow/metrics/_plots.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        )
+        precisions.append(precision)
+        recalls.append(recall)
+        f1s.append(f1)
+    return Curve(x=thresholds, y=f1s, label=label), Curve(x=recalls, y=precisions, label=label)
+
+
+def compute_object_detection_test_case_plots(
+    all_matches: List[MulticlassInferenceMatches],
+    test_case_name: str,
```
Comment (munkyshi) [inline kolena/workflow/metrics/_plots.py]: Is this test case specific, or just used for naming of the curves? It seems like we'd like these metrics to be as general as possible, and the more we can avoid prescribing the use case, the better (would `plot_name` or `curve_name` work just as well?).

File: kolena/workflow/metrics/__init__.py [hunk 3]
```diff
@@ -15,8 +15,10 @@
 from ._geometry import iou
 from ._geometry import match_inferences
 from ._geometry import match_inferences_multiclass
+from ._plots import compute_object_detection_test_case_plots
 
 __all__ = [
+    "compute_object_detection_test_case_plots",
```
Comment (munkyshi) [inline kolena/workflow/metrics/__init__.py:21]: Naming is a bit of a mouthful here, and doesn't tell a user what plots are actually being computed -- anything we can do to improve this?

`compute_pr_f1_plots`?

File: kolena/workflow/metrics/_plots.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+            y_score.append(bbox_inf.score)
+    return y_true, y_score, dict(arrays_by_label)
+
+
+def _compute_threshold_curves(
+    y_true: np.ndarray,
+    y_score: np.ndarray,
+    label: str,
+) -> Tuple[Curve, Curve]:
+    thresholds = list(np.linspace(min(abs(y_score)), max(y_score), min(401, len(y_score))))[:-1]
```
Comment (munkyshi) [inline kolena/workflow/metrics/_plots.py]: Curious about this piece:
```
min(401, len(y_score))
```
`len(y_score)` is strange to linspace with -- if you're going to have a point for every y_score, why not just use the threshold values (instead of linspacing)?

On the flip side, one consideration that we had for builtin OD was that linspacing gave us more points when values were sparse, but less detail than we would have liked when values were dense. Not a blocker, but worth pointing out that you could have 100 points between (0.9525, 0.9550) and only 10 points between (0, 0.5) -- linspacing smudges that out.

Last thing to point out -- there's a chance that you're missing the f1 optimal point if you linspace. Do we want to ensure that it gets surfaced?

**Review by munkyshi — COMMENTED** [review]
Haven't looked too deeply at the logic -- did we get a chance to compare this to outputs from existing OD plotting?

General question: Is the immediate use case for these metrics the generic OD implementation? Are we still planning on surfacing things like `F1-Optimal`?

nit: Can you add your name or initials to the branch you create for these? It helps with pruning / deleting old code (something like `mc/add-object-detection-test-case-plots` lets us know who's working on the branch / who to talk to if we want to clean up old branches)

File: kolena/workflow/metrics/_plots.py (hunk 4 above)
Comment (marko-polo-cheno) [reply kolena/workflow/metrics/_plots.py]: I made more progress on the entire workflow for a better understanding of the [full picture](https://github.com/kolenaIO/kolena/pull/56#discussion_r1214373355). Here, plot utils do not have to be at the test case level, but for the OD workflow, they will be per test case. I think a good solution to address missing the optimal threshold (for sparse cases) is doubling the size of the linspace.

Comment (marko-polo-cheno) [reply kolena/workflow/metrics/_plots.py]: @y27choi is doubling the size of the linspace from `min(401, len(y_score))` to `min(401, 2 * len(y_score))` a good idea?

Comment (y27choi) [reply kolena/workflow/metrics/_plots.py]: I'm not sure how doubling the size of the linspace can ensure including the f1 optimal threshold.  Can you explain?

In general, I agree with @munkyshi's comment.  I think we can definitely improve the way we plot PR / F1 curves, but I don't think it's a blocker.  Having said that, let's check popular implementations to see if there is a standard way of extracting the right amount of threshold data points.

And instead of doubling, what do you think of manually adding the f1 optimal threshold if it's not already included in `thresholds`?

File: kolena/workflow/test_sample.py
Comment (munkyshi) [inline kolena/workflow/test_sample.py:1]: Is this diff intended? It looks like it undoes some of the changes that @gordonhart recently put in.

File: pyproject.toml [hunk 5]
```diff
@@ -83,3 +83,6 @@ norecursedirs = "kolena"
 python_classes = []
 # only collect functions starting with 'test__'
 python_functions = ["test__*"]
+markers = [
+    "metrics: tests that extra dependencies for metrics",
```
Comment (munkyshi) [inline pyproject.toml:87]: This sentence feels incomplete.

File: tests/unit/workflow/test_evaluator.py
Comment (munkyshi) [inline tests/unit/workflow/test_evaluator.py:1]: Is this diff intended? It doesn't look like these tests are reintroduced anywhere.

File: kolena/_experimental/object_detection/utils.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+def _compute_sklearn_arrays(
+    all_matches: List[MulticlassInferenceMatches],
+) -> Tuple[List[int], List[int], Dict[str, List[int]], Dict[str, List[int]]]:
+    y_true_by_label: defaultdict[str, List[int]] = defaultdict(lambda: [])
+    y_score_by_label: defaultdict[str, List[int]] = defaultdict(lambda: [])
+    y_true: List[int] = []
+    y_score: List[int] = []
+    for image_bbox_matches in all_matches:
+        for _, bbox_inf in image_bbox_matches.matched:  # TN (if above threshold)
```
Comment (munkyshi) [inline kolena/_experimental/object_detection/utils.py:41]: nit: should comment be `TP`?

File: kolena/workflow/_validators.py
Comment (munkyshi) [inline kolena/workflow/_validators.py:1]: Also feels like this is an unintended diff?

File: kolena/workflow/evaluator.py
Comment (munkyshi) [inline kolena/workflow/evaluator.py:1]: Unintended diff?

File: kolena/workflow/metrics/__init__.py (hunk 3 above)
Comment (munkyshi) [reply kolena/workflow/metrics/__init__.py:21]: Did this get resolved?

**Comment by munkyshi** [comment]
Changes look good, just a few small comments, and it seems like an unintended diff got added.

File: kolena/_experimental/object_detection/utils.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+import numpy as np
+
+from kolena._extras.metrics.sklearn import sklearn_metrics
+from kolena.workflow.evaluator import Curve
+from kolena.workflow.evaluator import CurvePlot
+from kolena.workflow.metrics import MulticlassInferenceMatches
+
+
+def threshold_key(label: str) -> str:
```
Comment (y27choi) [inline kolena/_experimental/object_detection/utils.py:28]: Is this used anywhere?

File: kolena/_experimental/object_detection/utils.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    sanitized = re.sub(r"\W+", "_", label)
+    return f"threshold_{sanitized}"
+
+
+def _compute_sklearn_arrays(
+    all_matches: List[MulticlassInferenceMatches],
+) -> Tuple[List[int], List[int], Dict[str, List[int]], Dict[str, List[int]]]:
```
Comment (y27choi) [inline kolena/_experimental/object_detection/utils.py:35]: Don't scikit-learn arrays need to be `np.array` type?  Is there reason why we are using `List` instead of `np.array` here?

File: kolena/_experimental/object_detection/utils.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    y_true_by_label: defaultdict[str, List[int]] = defaultdict(lambda: [])
+    y_score_by_label: defaultdict[str, List[int]] = defaultdict(lambda: [])
```
Comment (y27choi) [inline kolena/_experimental/object_detection/utils.py:37]: shouldn't `y_score` be float?

File: kolena/_experimental/object_detection/utils.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    y_true: List[int] = []
+    y_score: List[int] = []
```
Comment (y27choi) [inline kolena/_experimental/object_detection/utils.py:39]: shouldn't `y_score` be float?

File: kolena/_experimental/object_detection/utils.py
Comment (y27choi) [inline kolena/_experimental/object_detection/utils.py:1]: Are these utility functions only for multi-class OD?  If so, what's the plan for single class OD?

File: kolena/_experimental/object_detection/utils.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+    y_true: np.ndarray,
+    y_score: np.ndarray,
+    label: str,
+) -> Tuple[Curve, Curve]:
+    thresholds = list(np.linspace(min(abs(y_score)), max(y_score), min(401, len(y_score))))[:-1]
+    precisions: List[float] = []
+    recalls: List[float] = []
+    f1s: List[float] = []
+    for threshold in thresholds:
+        y_pred = [1 if score > threshold else 0 for score in y_score]
```
Comment (y27choi) [inline kolena/_experimental/object_detection/utils.py:69]: Should it be counted as a positive prediction if `score > threshold` or `score >= threshold`?

File: kolena/_experimental/object_detection/utils.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+    return Curve(x=thresholds, y=f1s, label=label), Curve(x=recalls, y=precisions, label=label)
+
+
+def compute_pr_f1_plots(
+    all_matches: List[MulticlassInferenceMatches],
+    curve_label: str = "baseline",
+) -> List[CurvePlot]:
+    """
+    Creates a PR (precision and recall) curve and F1-threshold (confidence threshold) curve for the multiclass object
+    detection workflow. For `n` labels, each plot has `n+1` curves. One for the test case, and one per label.
```
Comment (y27choi) [inline kolena/_experimental/object_detection/utils.py:88]: Should update this now that we don't include per label plots.

File: kolena/_experimental/object_detection/utils.py [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+
+    :param all_matches: a list of multiclass or single class matching results.
```
Comment (y27choi) [inline kolena/_experimental/object_detection/utils.py:90]: Does it include `InferenceMatches`?

File: kolena/_experimental/object_detection/utils.py [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+    :param curve_label: the label of the main curve.
+    :return: Two :class:`CurvePlot`s for the PR curves and F1-threshold curves for the test case and each label.
```
Comment (y27choi) [inline kolena/_experimental/object_detection/utils.py:92]: Update here as well.

File: tests/unit/_experimental/test_plots.py [hunk 15]
```diff
@@ -0,0 +1,1229 @@
...
+    for c1, c2 in zip(c1.curves, c2.curves):
+        assert_curve_equal(c1, c2)
+
+
+@pytest.mark.metrics
+@pytest.mark.parametrize(
+    "test_name, matchings, f1_curve, pr_curve",
+    [
+        (
+            "zeros with unmatched gt and unmatched inf",
```
Comment (y27choi) [inline tests/unit/_experimental/test_plots.py:57]: This is an interesting case.  What happens if there are no valid matches?  Should we still return a plot with a single point at the origin?  @marko-polo-cheno do you know what it would look like on the webapp?

File: tests/unit/_experimental/test_plots.py [hunk 16]
```diff
@@ -0,0 +1,1229 @@
...
+                ],
+                x_config=None,
+                y_config=None,
+            ),
+            CurvePlot(
+                title="Precision vs. Recall",
+                x_label="Recall",
+                y_label="Precision",
+                curves=[
+                    Curve(x=[0.0, 0.0], y=[0.0, 0.0], label="zeros, but one match for label a"),
```
Comment (y27choi) [inline tests/unit/_experimental/test_plots.py:158]: Hmm shouldn't precision be `1.0` at threshold of `0`?  I guess this will depend on whether to include or exclude the matched instance right @ the threshold.

File: tests/unit/_experimental/test_plots.py [hunk 17]
```diff
@@ -0,0 +1,1229 @@
...
+                    ],
+                    unmatched_inf=[],
+                ),
+            ],
+            CurvePlot(
+                title="F1-Score vs. Confidence Threshold",
+                x_label="Confidence Threshold",
+                y_label="F1-Score",
+                curves=[
+                    Curve(x=[1.0], y=[0.0], label="zeros, but b is confused with a"),
```
Comment (y27choi) [inline tests/unit/_experimental/test_plots.py:216]: I'm curious why `x = 1.0` here but all the previous cases have `x = 0.0`.

File: tests/unit/_experimental/test_plots.py [hunk 18]
```diff
@@ -0,0 +1,1229 @@
...
+                    ],
+                    unmatched_inf=[],
+                ),
+            ],
+            CurvePlot(
+                title="F1-Score vs. Confidence Threshold",
+                x_label="Confidence Threshold",
+                y_label="F1-Score",
+                curves=[
+                    Curve(x=[0.0, 0.0, 0.0], y=[0.0, 0.0, 0.0], label="no confusion, one TP per label"),
```
Comment (y27choi) [inline tests/unit/_experimental/test_plots.py:285]: nit: probably doesn't really matter, but I think `np.linspace(0, 0, 4)[:-1]` results in duplicate thresholds of `0` — this is why we see `(4-1)` `(0, 0)` points.  Should we drop dup threshold?

File: tests/unit/_experimental/test_plots.py [hunk 19]
```diff
@@ -0,0 +1,1229 @@
...
+                            ScoredLabeledBoundingBox((10.0, 10.0), (20.0, 20.0), "cow", 0.9),
+                        ),
+                    ],
+                    unmatched_gt=[
+                        (
+                            LabeledPolygon(
+                                points=[(1, 1), (1, 2), (2, 2), (2, 1)],
+                                label="cow",
+                            ),
+                            ScoredLabeledPolygon(
```
Comment (y27choi) [inline tests/unit/_experimental/test_plots.py:601]: Shouldn't this be also added to `unmatched_inf`?  
Currently this pair is only counted as 1 FN, but shouldn't this be 1 FN and 1 FP?

**Review by y27choi — COMMENTED** [review]
Left some comments.  I will also double check the numbers in the unit tests.

**Review by y27choi — COMMENTED** [review]
Left some comments.  I will also double check the numbers in the unit tests.

File: kolena/_experimental/object_detection/utils.py (hunk 8 above)
Comment (gordonhart) [reply kolena/_experimental/object_detection/utils.py:35]: To add to this, if using NumPy arrays the type hint should be `np.ndarray`

File: tests/unit/_experimental/test_plots.py (hunk 18 above)
Comment (y27choi) [reply tests/unit/_experimental/test_plots.py:285]: It seems like sklearn uses only the unique thresholds.  [sklearn link](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.precision_recall_curve.html)
image

File: tests/unit/_experimental/test_plots.py [hunk 20]
```diff
@@ -0,0 +1,1229 @@
...
+                    ],
+                ),
+            ],
+            CurvePlot(
+                title="F1-Score vs. Confidence Threshold",
+                x_label="Confidence Threshold",
+                y_label="F1-Score",
+                curves=[
+                    Curve(
+                        x=[
```
Comment (y27choi) [inline tests/unit/_experimental/test_plots.py:1020]: The current implementation produces `len(y_score)` (when there are < 401), but here `y_score` includes TP, FP, FN.  Does this make sense?  FNs don't have any valid scores (i.e. we are using `-1` here) so shouldn't there be only # of TP + FP thresholds?

File: tests/unit/_experimental/test_plots.py [hunk 21]
```diff
@@ -0,0 +1,1229 @@
...
+                x_config=None,
+                y_config=None,
+            ),
+            CurvePlot(
+                title="Precision vs. Recall",
+                x_label="Recall",
+                y_label="Precision",
+                curves=[
+                    Curve(
+                        x=[
```
Comment (y27choi) [inline tests/unit/_experimental/test_plots.py:1120]: I tried running sklearn's `precision_recall_curve` and compared the results. 
It seems like precision and recall are incorrect @ threshold of `0.01`.  Could you please take a look at this @marko-polo-cheno ?

[pr_curve.py.zip](https://github.com/kolenaIO/kolena/files/11682132/pr_curve.py.zip)

File: kolena/_experimental/object_detection/utils.py
Comment (marko-polo-cheno) [reply kolena/_experimental/object_detection/utils.py:1]: Will add `InferenceMatches` for pr and f1 plots

File: tests/unit/_experimental/test_plots.py (hunk 21 above)
Comment (marko-polo-cheno) [reply tests/unit/_experimental/test_plots.py:1120]: In the new PR, I think I've addressed all the comments. Most are due to a logic change. This case now matches perfectly, with the exception of the threshold @ -1, which I omit.

The two naive but massive issues are non-distinct thresholds with the `>=` vs `>` choice.
